osquery/filesystem/filesystem.cpp

```diff
@@ -294,25 +323,49 @@ Status resolveLastPathComponent(const boost::filesystem::path& fs_path,
         processed_path + "/" +
         components[components.size() - 1].substr(
             0, components[components.size() - 1].find(kWildcardCharacter, 1));
-    for (const auto& file : files) {
-      if (file.find(prefix, 0) != 0) {
-        continue;
+    if (setting & REC_LIST_FOLDERS) {
+      for (const auto& fold : folders) {
+        if (fold.find(prefix, 0) != 0) {
+          continue;
+        }
+        results.push_back(fold);
+      }
+    }
+    if (setting & REC_LIST_FILES || setting & REC_EVENT_OPT) {
+      for (const auto& file : files) {
+        if (file.find(prefix, 0) != 0) {
+          continue;
+        }
+        results.push_back(file);
       }
-      results.push_back(file);
     }
+    //Should be a return here?
+    return Status(0, "OK");
   }
 
   // Is this a %(.*) type file match
   if (components[components.size() - 1][0] == kWildcardCharacter[0]) {
     std::string suffix = components[components.size() - 1].substr(1);
-
-    for (const auto& file : files) {
-      boost::filesystem::path p(file);
-      std::string file_name = p.filename().string();
-      size_t pos = file_name.find(suffix);
-      if (pos != std::string::npos &&
-          pos + suffix.length() == file_name.length()) {
-        results.push_back(file);
+    if (setting & REC_LIST_FOLDERS) {
+      for (const auto& fold : folders) {
+        boost::filesystem::path p(fold);
+        std::string file_name = p.filename().string();
```

Comment:
If you don't need a reference to `p` then `boost::filesystem::path(fold).filename().string()` works too.